### ISSUE TITLE
`special_module_name` triggered for a directory module

### ISSUE THREAD

**Issue by Tuetuopay** [rust-lang/rust#104161]

**Comment by correabuscar** [comment]
I encounter this warning when I try to have the lib inside the bin project, aka the `lib.rs` and the `main.rs` in `src/`, like the following:


`$ cargo new dual_bin_and_lib_special_module_name`
Use this `src/main.rs` contents:
```rust
mod lib;

fn main() {
    println!("Hello, world!{}", lib::add(1,2));
}
```
Make this file `src/lib.rs`:
```rust
pub fn add(left: usize, right: usize) -> usize {
    left + right
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn it_works() {
        let result = add(2, 2);
        assert_eq!(result, 4);
    }
}
```

`$ cargo run`:
```rust
$ cargo run
   Compiling dual_bin_and_lib_special_module_name v0.1.0 (/home/user/sandbox/rust/05_sandbox/dynamic_lib/dual_bin_and_lib_special_module_name)
warning: found module declaration for lib.rs
 --> /home/user/sandbox/rust/05_sandbox/dynamic_lib/dual_bin_and_lib_special_module_name/src/main.rs:1:1
  |
1 | mod lib;
  | ^^^^^^^^
  |
  = note: lib.rs is the root of this crate's library target
  = help: to refer to it from other targets, use the library's name as the path
  = note: `#[warn(special_module_name)]` on by default

warning: `dual_bin_and_lib_special_module_name` (bin "dual_bin_and_lib_special_module_name") generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.20s
     Running `target/debug/dual_bin_and_lib_special_module_name`
Hello, world!3
```

--------------

To workaround it, I've found this is needed:
```diff
diff -upr dual_bin_and_lib_special_module_name/Cargo.toml dual_bin_and_lib_special_module_name_fixed/Cargo.toml
--- dual_bin_and_lib_special_module_name/Cargo.toml	2024-04-05 22:57:08.746022868 +0200
+++ dual_bin_and_lib_special_module_name_fixed/Cargo.toml	2024-04-05 22:56:58.386023284 +0200
@@ -4,5 +4,7 @@ version = "0.1.0"
 edition = "2021"
 
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+[lib]
+path = "src/lib2.rs"
 
 [dependencies]
Only in dual_bin_and_lib_special_module_name_fixed/src: lib2.rs
Only in dual_bin_and_lib_special_module_name/src: lib.rs
diff -upr dual_bin_and_lib_special_module_name/src/main.rs dual_bin_and_lib_special_module_name_fixed/src/main.rs
--- dual_bin_and_lib_special_module_name/src/main.rs	2024-04-05 22:57:18.569355835 +0200
+++ dual_bin_and_lib_special_module_name_fixed/src/main.rs	2024-04-05 22:56:58.386023284 +0200
@@ -1,5 +1,5 @@
-mod lib;
+mod lib2;
 
 fn main() {
-    println!("Hello, world!{}", lib::add(1,2));
+    println!("Hello, world!{}", lib2::add(1,2));
 }
```
`cargo run`
```rust
$ cargo run
   Compiling dual_bin_and_lib_special_module_name v0.1.0 (/home/user/sandbox/rust/05_sandbox/dynamic_lib/dual_bin_and_lib_special_module_name_fixed)
    Finished dev [unoptimized + debuginfo] target(s) in 0.19s
     Running `target/debug/dual_bin_and_lib_special_module_name`
Hello, world!3
```

Is there a better way ?
Thanks.

**Comment by Alexendoo** [comment]
This response is a *little* late but it may help others

For @correabuscar's example the intention would be for you to remove `mod lib` and refer to it as `dual_bin_and_lib_special_module_name::add` rather than `lib::add`

**Comment by correabuscar** [comment]
> This response is a _little_ late but it may help others
> 
> For @correabuscar's example the intention would be for you to remove `mod lib` and refer to it as `dual_bin_and_lib_special_module_name::add` rather than `lib::add`

Thanks, that works and I now understand what the warning meant:
>   = note: lib.rs is the root of this crate's library target
  = help: to refer to it from other targets, **use the library's name as the path**
  = note: `#[warn(special_module_name)]` on by default
